Parse SAS format strings for XPT outputs (#257)

gorcha opened the pull request:
This PR adds a simple parser to extract format name, width and decimals from a SAS format string to be used when writing XPT files.

It parses the format string in reverse, assigns width and decimals based on the pattern of numbers and periods at the end, and uses the remainder of the string as the format name.

There are also a couple of tweaks to the section that writes the long format name for XPT v8 files:
* Now uses a simplified version of the parser to extract the format name
* Writes the entire format name to the long format record per the [spec](https://support.sas.com/content/dam/SAS/support/en/technical-papers/record-layout-of-a-sas-version-8-or-9-data-set-in-sas-transport-format.pdf)
* The long format record was incorrectly writing the label out after the format/informat values, these have been swapped.

gorcha commented:
I've made a few more updates after seeing the test failures. The XPT reader code has been updated to reflect the changes to the XPT writer.

I had to make a couple of changes to the tests for what looked like false positives, and I think they make sense:
* One of the SAS format roundtrip tests has been changed to use `$CHAR3` instead of `$CHAR3.`. The trailing period is dropped when roundtripping XPT files because of the way the format is reconstructed so `$CHAR3.` is read back in as `$CHAR3` and fails the test. It represents the same format so it made to sense to me to update the test.
* It now checks if label_set has been set before running the roundtrip test (as was already the case for format and display_width). It was throwing some strange errors in the display_width tests that looked like it was checking unallocated memory like:
```
Test "Display width" failed: Column label sets
 * Format: xpt5 (0x200000)
 * Expected: "¤~y²U" (length=6)
 * Received: "~y²U" (length=5)
 * Column: 1
```

evanmiller commented on src/sas/readstat_xport_write.c:
```diff
@@ -173,31 +224,28 @@ static readstat_error_t xport_write_variables(readstat_writer_t *writer) {
+                int j;
+                for (j=format_len-1; j>=0; j--) {
+                  if (!((variable->format[j] >= 48 && variable->format[j] <= 57) ||
+                          variable->format[j] == 46)) {
```
Maybe reuse the parser from above? If not, please use the C literal syntax for characters, `'0'`, `'9'`, etc

gorcha replied:
Yep I think best to use the parser from above once it's set up. Thanks!

gorcha commented on src/sas/readstat_xport_write.c:
```diff
@@ -109,9 +109,60 @@ static readstat_error_t xport_write_variables(readstat_writer_t *writer) {
+             * 3 - second number (width)
+             * 4 - name
+             */
+            int state = 0;
```
Shall do, thanks!

gorcha commented:
I've updated the format parsing code to use a Ragel parser and added an xport_format_t struct to store the parser output, and added a few more tests to check it's throwing errors correctly.

I've also had a go at the fuzzer setup but a bit less sure of what I'm doing there!

evanmiller commented on Makefile.am:
```diff
@@ -95,6 +96,7 @@ noinst_HEADERS = \
        src/sas/readstat_sas.h \
        src/sas/readstat_sas_rle.h \
        src/sas/readstat_xport.h \
+	   src/sas/readstat_xport_parse_format.h \
```
Please fix the whitespace here

gorcha replied:
Done!

evanmiller reviewed (changes requested):
Looking good! Just found one whitespace issue, otherwise I'm happy with it.

Thanks for taking the time to implement the fuzzer – and of course the word dictionary is a nice touch.

evanmiller commented:
It looks like Appveyor is failing:

https://ci.appveyor.com/project/evanmiller/readstat/builds/41638815

You may need to modify the Visual Studio project file to ensure all the new files are being built. See

https://github.com/WizardMac/ReadStat/blob/master/VS17/ReadStat.vcxproj

evanmiller commented:
Thanks, much better now!

mstackhouse commented:
@evanmiller - any idea when these changes might be able to be merged into a new release of ReadStat? These changes would open up some [important doors](https://github.com/tidyverse/haven/issues/650) for haven, which will have a significant impact in some pharmaceutical industry projects.
